internal/storage/storage.go

```diff
@@ -0,0 +1,220 @@
+	// or UID mismatch.
+	ErrConflict = errors.New("operation failed because of a Version or Uid mismatch")
+
+	// ErrInconsistent indicates that the attempted write or consistent read could
+	// not be achieved because of a consistency or availability issue (e.g. loss of
+	// quorum, or when interacting with a Raft follower).
+	ErrInconsistent = errors.New("cannot satisfy consistency requirements")
+)
+
+// Backend provides the low-level storage substrate for resources. It can be
+// implemented using internal (i.e. Raft+MemDB) or external (e.g. DynamoDB)
+// storage systems.
+//
+// Refer to the method comments for details of the behaviors and invariants
+// provided, which are also verified by the conformance test suite in the
+// internal/storage/conformance package.
+//
+// Cross-cutting concerns:
+//
+// # UIDs
+//
+// Users identify resources with a name of their choice (e.g. service "billing")
+// but internally, we add our own identifier in the Uid field to disambiguate
+// references when resources are deleted and re-created with the same name.
+//
+// # GroupVersion
+//
+// In order to support automatic translation between schema versions, we only
+// store a single version of a resource, and treat types with the same Group
+// and Kind, but different GroupVersions, as equivalent.
+type Backend interface {
+	// Read a resource using its ID.
+	//
+	// # UIDs
+	//
+	// If id.Uid is empty, Read will ignore it and return whatever resource is
+	// stored with the given name. This is the desired behavior for user-initiated
+	// reads.
+	//
+	// If id.Uid is non-empty, Read will only return a resource if its Uid matches,
+	// otherwise it'll return ErrNotFound. This is the desired behaviour for reads
+	// initiated by controllers, which tend to operate on a specific lifetime of a
+	// resource.
+	//
+	// See Backend docs for more details.
+	//
+	// # GroupVersion
+	//
+	// If id.Type.GroupVersion doesn't match what is stored, Read will return a
+	// GroupVersionMismatchError which contains a pointer to the stored resource.
+	//
+	// See Backend docs for more details.
+	//
+	// # Consistency
+	//
+	// Read makes no guarantees about consistency, and may return stale results.
+	// For stronger guarantees, use ReadConsistent.
+	Read(ctx context.Context, id *pbresource.ID) (*pbresource.Resource, error)
+
+	// ReadConsistent provides the same functionality as Read, but guarantees
+	// single-resource sequential consistency, typically by bypassing any caches
+	// and proxying the request directly to the underlying storage system.
+	//
+	// If a consistent read cannot be achieved (e.g. when interacting with a Raft
+	// follower, or quorum is lost) ErrInconsistent will be returned.
+	//
+	// Use ReadConsistent sparingly, and prefer Read when possible.
+	ReadConsistent(ctx context.Context, id *pbresource.ID) (*pbresource.Resource, error)
+
+	// WriteCAS performs an atomic CAS (Check-And-Set) write of a resource based
```

Comment:
[nit/question] I seems like [consul docs for kv put](https://developer.hashicorp.com/consul/commands/kv/put) refer to `-cas` as a Check-And-Set operation, but I think everyone else refers to CAS as Compare And Swap.  Our implementation is consistent with Compare and Swap.  I think it might be confusing since Test-And-Set is another locking operation type.  If this is all inadvertent, I'd like to change this here and in another PR, we can update the consul docs.  I'm just not sure if I am missing a broader use of `Check-And-Set`.

Reply:
Ah yes, I did copy it from the KV definition of CAS. Let's go with Compare-And-Swap, as it's more accurately descriptive.

Reply:
Fixed in https://github.com/hashicorp/consul/pull/16538/commits/f54bcb8024ab0b3176850287120799df004cbaf3